chore: improve OLM environment detection

sxd opened the pull request:
As far as now, we were relying on the detection of the SecurityContextConstraint to know that we were running on an OLM environment but this is not a secure way since an OLM environment may or may not have a SCC inside, that's why we should look at the `operators.coreos.com/v1` resources and check that  it exists, if this is the case, we're in an OLM environment.

This helps us to improve the plugin report and properly retrieve more useful information.

Closes #3764

sxd committed:
chore: improve OLM environment detection

As far as now, we were relying on the detection of the SecurityContextConstraint
to know that we were running on an OLM environment but this is not a secure way
since an OLM environment may or may not have a SCC inside, that's why we should
look at the `operators.coreos.com/v1` resources and check that  it exists, if this
is the case, we're in an OLM environment.

This helps us to improve the plugin report and properly retrieve more useful
information.

Closes #3764

Signed-off-by: Jonathan Gonzalez V <jonathan.gonzalez@enterprisedb.com>

armru committed:
chore: lint

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
chore: simplify labels

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru committed:
fix: merge list and review

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

fcanovai committed:
fix: review

Signed-off-by: Francesco Canovai <francesco.canovai@enterprisedb.com>

fcanovai commented on internal/cmd/plugin/report/olm.go:
```diff
@@ -0,0 +1,120 @@
+		return nil, fmt.Errorf("could note get resource: %v, %v", gvr, err)
+	}
+
+	for idx := range listFromLabelOpenshift.Items {
+		item := listFromLabelOpenshift.Items[idx]
+		if !isContained(item, listFromLabelNamespace.Items) {
+			listFromLabelOpenshift.Items = append(listFromLabelOpenshift.Items, item)
+		}
+	}
+
+	return listFromLabelOpenshift, err
```
This won't merge the lists. If the operator isn't in `openshift-operators`, but we have an operator in a namespace, we return empty list.

fcanovai commented on internal/cmd/plugin/report/olm.go:
```diff
@@ -0,0 +1,120 @@
+	return listFromLabelOpenshift, err
+}
+
+// olmOperatorReport contains the operator data in Oolm
```
```suggestion
// olmOperatorReport contains the operator data in olm
```

fcanovai reviewed (changes requested):
Beside the broken piece of code, I'm not convinced about the entire logic.

This PR is checking the `openshift-operators` namespace, and the namespace defined in the plugin call (or `default`). However, it first checks for the operator deployment in the plugin-defined namespace, exiting if the operator is not there. So looking into openshift-operator is not useful, you have to get the operator location right. I think we could just check the namespace asked for in the plugin.

fcanovai committed:
fix: plugin-defined namespace

Remove broken detection logic. We check for the operator in the
namespace the user provided.

Signed-off-by: Francesco Canovai <francesco.canovai@enterprisedb.com>

armru commented on internal/cmd/plugin/report/olm.go:
```diff
@@ -0,0 +1,98 @@
+	ctx context.Context,
+	dynamicClient dynamic.Interface,
+	namespace, resource string,
+) (client.ObjectList, error) {
```
Wouldn't make it better to return a []client.object instead of an ObjectList?

sxd committed:
chore: review

Signed-off-by: Jonathan Gonzalez V. <jonathan.gonzalez@enterprisedb.com>

fcanovai commented:
/ok-to-merge
